Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp

```diff
@@ -39,12 +39,110 @@ GStreamerVideoCapturer::GStreamerVideoCapturer(const char* sourceFactory)
 GstElement* GStreamerVideoCapturer::createConverter()
 {
     auto converter = gst_parse_bin_from_description("videoscale ! videoconvert ! videorate", TRUE, nullptr);
+#if USE(WESTEROS_SINK)
+// If camera supports only x-raw, westerossink expects encoded form. So pass through an h264 encoder and parser after probing platform.
+// If camera supports encoded formats, use a dummy identity element to skip converting to x-raw.
+    const char* codec = getPreferredCodec().c_str();
+    GST_INFO_OBJECT(m_pipeline.get(), "Preferred codec is %s ", codec);
+
+    if (strcmp (codec, "video/x-raw") == 0)
+    {
+        GstElementFactory *h264encoder = getEncoder ("video/x-h264");
+        GstElementFactory *h264parser = getParser ("video/x-h264");
+
+        char *videoInput = g_strdup_printf ("videoscale ! videoconvert ! videorate ! %s ! %s", GST_ELEMENT_NAME(h264encoder), GST_ELEMENT_NAME(h264parser));
+        converter = gst_parse_bin_from_description(videoInput, TRUE, nullptr);
+        m_caps = adoptGRef(gst_caps_new_empty_simple("video/x-h264"));
+    }
+    else
+    {
+        converter = gst_parse_bin_from_description("identity", TRUE, nullptr);
+        m_caps = adoptGRef(gst_caps_new_empty_simple(codec));
+    }
+#endif
 
     ASSERT(converter);
 
     return converter;
 }
 
+std::string GStreamerVideoCapturer::getPreferredCodec()
+{
+    GstCaps* caps = gst_device_get_caps(m_device.get());
+    const std::string videoCaps = gst_caps_to_string(caps);
```

Comment:
`gst_caps_to_string()` returns a newly allocated string ([the function is transfer-full](https://gstreamer.freedesktop.org/documentation/gstreamer/gstcaps.html?gi-language=c#gst_caps_to_string)), which will be copied into the `std::string` and forgotten (leaked) after that.
If you really prefer to have a `std::string` in order to use `find()` method, I'd suggest something like this:
```
const std::string videoCaps = GUniquePtr<gchar>(gst_caps_to_string(caps)).get();
```
`GUniquePtr` will take care of freeing the `char*` when the execution of the source code line finishes. `videoCaps` will hold a copy and will be destroyed (and the copy freed) when the execution flow goes out of the function.
You can also make `videoCaps` to be a GUniquePtr and use `!g_strstr_len(videoCaps.get(), -1, "video/x-av1")` to do the comparisons. See [gst_caps_to_string()](https://developer.gnome.org/glib/stable/glib-String-Utility-Functions.html#g-strstr-len) for more info.

Reply:
Changed